splink/input_column.py

```diff
@@ -1,112 +1,153 @@
-        identifier.args["quoted"] = False
-    return tree
+    def add_column_reference(self, index_name: str | int):
+        """
+        Manually add a column reference to your column - ['idx'], for example.
+
+        This is useful during the parsing process or when you try to combine
+        illegal name and bracket combinations such as 'SUR NAME['lat']',
+        without needing to add quotes to the column identifier.
+        """
+        is_str = type(index_name) == str
+        self.column_reference = exp.Literal(this=index_name, is_string=is_str)
+        return self
+
+    def build_column_tree(self):
+        # Create our column representation
+        column = sqlglot.column(col=self.name, table=self.table, quoted=self.quoted)
+        # Column is made up of the identifier + bracket index (if it exists)
+        if self.column_reference:
+            return exp.Bracket(this=column, expressions=[self.column_reference])
+        else:
+            return column
+
+    @classmethod
+    def from_sqlglot_column(
+        cls, column_tree: exp.Column, column_reference: exp.Identifier = None
+    ):
+        return cls(column_tree.name, column_reference=column_reference)
+
+    def __repr__(self):
+        return (
+            f"{self.__class__.__name__}("
+            f"name='{self.name}', quoted={self.quoted}, "
+            f"column_reference={self.column_reference.sql()}"
+        )
 
 
 class InputColumn:
     """
     Represents a SQL column or column reference
     Handles SQL dialect-specific issues such as identifier quoting.
 
-    The input should be the raw identifier, without SQL-specific identifier quotes.
-
-    For example, if a column is named 'first name' (with a space), the input should be
-    'first name', not '"first name"'.
+    The input can be either the raw identifier, or an identifier with
+    SQL-specific identifier quotes.
 
     Examples of valid inputs include:
     - 'first_name'
-    - 'first name'
+    - 'first[name'
+    - '"first name"'
     - 'coordinates['lat']'
+    - '"sur NAME"['lat']
     - 'coordinates[1]'
-
     """
 
     def __init__(
-        self, raw_column_name_or_column_reference, settings_obj=None, sql_dialect=None
+        self,
+        raw_column_name_or_column_reference: str,
+        settings_obj=None,
+        sql_dialect: str = None,
     ):
         # If settings_obj is None, then default values will be used
         # from the jsonschama
         self._settings_obj = settings_obj
 
-        if sql_dialect:
-            self._sql_dialect = sql_dialect
-        elif settings_obj:
-            self._sql_dialect = self._settings_obj._sql_dialect
-        else:
-            self._sql_dialect = None
+        self.register_dialect(sql_dialect)
 
-        self.input_name = self._quote_if_sql_keyword(
+        self.input_name: str = self._quote_if_sql_keyword(
             raw_column_name_or_column_reference
         )
+        self.column_tree_builder: ColumnTreeBuilder = (
```

Comment:
I think the critical thing is that this 'base' column is immutable within this class

For example, you can ask for `first_name_l`, but the `first_name` always needs to be 'remembered' as the base information that's going to be manipulated.  With that in mind, I think this should be called something like the `base_column_tree`